vesting/lib.rs

```diff
@@ -0,0 +1,273 @@
+        pub fn releasable_balance(&self) -> Balance {
+            return self.vested_amount() as Balance - self.released_balance();
+        }
+
+        /// This calculates the amount that has already vested
+        /// but hasn't been released from the contract yet.
+        #[ink(message)]
+        pub fn vested_amount(&self) -> Balance {
+            return self.vesting_schedule(self.this_contract_balance(), self.time_now());
+        }
+
+        /// This sends the releasable balance to the beneficiary.
+        /// wallet address; no matter who triggers the release.
+        #[ink(message)]
+        pub fn release(&mut self) -> Result<(), Error> {
+            let releasable = self.releasable_balance();
+            if releasable == 0 {
+                return Err(Error::ZeroReleasableBalance);
+            }
+
+            self.released_balance += releasable;
+            self.env()
+                .transfer(self.beneficiary, releasable)
+                .expect("Transfer failed during release");
+
+            self.env().emit_event(Released {
+                value: releasable,
+                to: self.beneficiary,
+            });
+
+            Ok(())
+        }
+
+        /// This calculates the amount of tokens that have vested up
+        /// to the given current_time.
+        ///
+        /// The vesting schedule is linear, meaning tokens are
+        /// released evenly over the vesting duration.
+        ///
+        /// # Parameters:
+        ///     - total_allocation: The total number of tokens
+        ///       allocated for vesting.
+        ///     - current_time: The current timestamp for which
+        ///       we want to check the vested amount.
+        ///
+        /// # Returns:
+        ///     - `0` if the current_time is before the vesting start time.
+        ///     - total_allocation if the current_time is after the vesting
+        ///       end time or at least equal to it.
+        ///     - A prorated amount based on how much time has passed since
+        ///       the start of the vesting period if the `current_time` is
+        ///       during the vesting period.
+        ///     
+        /// # Example:
+        /// If the vesting duration is 200 seconds and 100 seconds have
+        /// passed since the start time, then 50% of the total_allocation
+        /// would have vested.
+        ///
+        pub fn vesting_schedule(
+            &self,
+            total_allocation: Balance,
+            current_time: Timestamp,
+        ) -> Balance {
+            if current_time < self.start_time() {
+                return 0;
+            } else if current_time >= self.end_time() {
+                return total_allocation;
+            } else {
+                return (total_allocation
+                    * (current_time - self.start_time()) as Balance)
+                    / self.duration_time() as Balance;
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
```

Comment:
Can you please also add a couple of E2E tests?

Reply:
Thanks for the feedback, I will proceed with this request today evening

Reply:
> Can you please also add a couple of E2E tests?

I'm noticing that the substrate-contracts-node doesn't yet offer an interval mining feature similar to Hardhat for eth chain development; this makes E2E testing challenging for time-based contracts, as is the case here.
Once this feature is available, I'm open to revisiting this. I could then submit an additional pull request to enhance this contract with a few E2E tests.
For now, the current unit tests accurately reflecting the contract's behavior on testnets, such as Rococo and Aleph Zero, ect...

Please let me know what's your take on this.